Show normal user popover for bot owners instead of extended profile.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This PR changes the user popover for bot owner to be normal one instead of extended profile one.

This also has two prep commits-
1. Rename show_user_info_popover to show_sender_info_popover as this is used to show the user popover for message sender.
(A function named show_user_info_popover is added which is used for bot owners case and also will be used for 'invited_by' column in invites table.)
2.  Change the popover to show 'Reply mentioning user' option only in msg sender popover and not on popovers openend from right sidebar.

**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![right](https://user-images.githubusercontent.com/35494118/85790784-72d33d80-b74e-11ea-897e-01c0e441a0cd.gif)
<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

clarammdantas commented on static/js/overlays.js:
```diff
@@ -134,6 +134,8 @@ exports.open_modal = function (selector) {
 };
 
 exports.close_overlay = function (name) {
+    popovers.hide_all();
```
This feels like it should be done in another commit.

sahil839 replied:
Agreed. I have made separate commit and also explained the need to add this in the commit message clearly. Thanks.

clarammdantas commented on static/templates/user_info_popover_content.hbs:
```diff
@@ -111,11 +111,13 @@
-            <i class="fa fa-at" aria-hidden="true"></i> {{#tr this}}Reply mentioning user{{/tr}} {{#if is_sender_popover}}<span class="hotkey-hint">(@)</span>{{/if}}
-        </a>
-    </li>
+        {{#if is_sender_popover}}
```
We should avoid adding logic to the templates. Could you try to that on the `.js` corresponding file instead?

sahil839 replied:
We handle all conditions for showing/hiding different options in the popovers in templates itself. We do not do it in popover.js for any popover, so I think it should be correct to do it here to maintain consistency.

timabbott replied:
Yeah, I think the template is the best place for this conditional.

timabbott replied:
And we should rename this conditional to `has_message_context`; that's more correct for what it means -- if there's an associated message to reply to or not.

clarammdantas commented:
Thanks for working on this PR, @sahil839! This is a really nice change.

I posted some comments on the code :)

timabbott commented on static/js/popovers.js:
```diff
@@ -744,7 +744,7 @@ exports.register_click_handlers = function () {
             user = people.get_by_email(email);
         }
-        show_user_info_popover(this, user, message);
+        show_sender_info_popover(this, user, message);
```
I don't think this rename is accurate; the code path here is clicking on a mention of the user.  I'd prefer to just skip this commit.

sahil839 replied:
I had rename this because for bot owners case we have no related message. 
I agree that renaming to 'show_sender_info_popover` is not correct as it is also used for mentions.
But I think we still need to rename it to keep it separate from bot owners and invites case(which will be added soon)  and I am thinking of other names, do you have any suggestions for for the name?

sahil839 replied:
Till now, I have thought of `show_user_info_popover_for_message`, but I do think we can think of something better. Any suggestions from your side?

sahil839 commented:
@timabbott I have added the commit for showing name of referrer instead of email and showing user-popover on clicking the name. 
And regarding the rename of show_user_info_popover, I have renamed it to show_user_info_popover_for_message for now, I will change it if you have any other suggestions.

timabbott commented on frontend_tests/node_tests/popovers.js:
```diff
@@ -171,6 +171,7 @@ run_test("sender_hover", (override) => {
                     is_active: true,
                     is_bot: undefined,
                     is_sender_popover: true,
+                    has_message_context: true,
```
I think we actually do want "Reply mentioning user" in the right sidebar popovers, though -- it's useful when you find a user in the right sidebar and want to generate mention syntax for them.  I suppose we could rename to "Copy mention syntax" when `!has_message_context` to address that use case without the weird connection to a message that's selected in a totally different part of the UI?

timabbott commented on static/js/settings_panel_menu.js:
```diff
@@ -19,6 +19,7 @@ class SettingsPanelMenu {
 
         this.main_elem.on("click", "li[data-section]", (e) => {
             const section = $(e.currentTarget).attr("data-section");
+            popovers.hide_all();
```
This sort of change doesn't make sense to put in a click handler -- since then if we add some other UI widget that can change the view, it won't run.  Does this belong in `activate_section_or_default` instead?

timabbott commented on static/js/popovers.js:
```diff
@@ -347,6 +348,27 @@ exports.show_user_profile = function (user) {
+        return;
+    }
+    const elt = $(element);
+    if (elt.data("popover") === undefined) {
```
Is there a situation where we expect this to ever be false?

sahil839 replied:
No, there is no situation for this to be false, I have removed this condition.

sahil839 commented:
Updated the PR with the changes suggested above. I did not notice the bot owner mentioned in bot popover in messages. Have updated that also, but I am not sure how it looks, attaching the GIF below-
![bot-owner](https://user-images.githubusercontent.com/35494118/89667264-759e8200-d8f9-11ea-9eea-657dfc49df2c.gif)

timabbott commented:
Does the `Copy mention syntax` option do anything?  I think we need to actually implement it; it should work like "Copy link to conversation" in the message actions popover, I think?

I merged 1ae00a306b04b3189696863bb312a4e4d6529c8b and e1636d30a996833d2a3081b08a36b687edcf75ae since they look good.

sahil839 commented:
Currently `Copy mention syntax` just pastes the syntax in compose box and the stream and topic inputs are filled according to narrow. Do we want to change it to work like "Copy link to conversation"?

sahil839 commented:
@timabbott Updated the PR. 
Now "Copy mention syntax" works like "Copy link to conversation" and just copies the mention syntax to clipboard.

timabbott commented:
The "Copy user mention syntax" buttons here don't visibly do anything when I tested them.

timabbott commented:
I think the use of different attribute names (e.g. `data-referrer-id`) for each `view_user_profile` widget is a bit strange; do we need to do that or can we just have them all have a `data-user-id` attribute?  That would make #16361 interact with this transparently.

sahil839 commented:
@timabbott Updated the PR. 
- Added a prep commit for renaming keyboard handling functions for user popovers opened from message. 
- Added commit for using one attribute data-user-id and a single click handler for opening popovers.
- Added a prep commit to support closing of popovers in settings overlay using escape key. 
- Added keyboard support for user info popovers in a separate commit, so that it can be reviewed easily(can squash it with the main commit if needed).

timabbott commented:
Merged as 2381e25438d9feb087c7bc677100e4814e5c9c3a, after a few small tweaks (mainly to match of both name and email in the last commit), thanks @sahil839!  This took a lot of revision but is important preparation for other features.

If you're up for it, a good quick follow-up is https://github.com/zulip/zulip/issues/16359, which is an important pain point for a lot of open organizations.

@ryanreh99 FYI.

sahil839 commented:
I am up for working on #16359 but @shanukun has started work on it.
